File: ScanbotSDKDemo/LicensePlateDemoViewController.m
//
//  LicensePlateDemoViewController.m
//  SBSDK Internal Demo
//
//  Created by Sebastian Husche on 03.03.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

#import "LicensePlateDemoViewController.h"
@import ScanbotSDK;

@interface LicensePlateDemoViewController()<SBSDKLicensePlateScannerViewControllerDelegate>

@property(nonatomic, strong) SBSDKLicensePlateScannerViewController *scannerViewController;
@property (nonatomic, strong) IBOutlet UIView *cameraContainer;
@property (nonatomic, strong) IBOutlet UILabel *resultLabel;
@property (nonatomic, strong) IBOutlet UIImageView *resultImageView;
@end



@implementation LicensePlateDemoViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.resultImageView.layer.cornerRadius = 8.0;
    self.resultImageView.layer.masksToBounds = YES;
    self.resultImageView.layer.borderColor = [UIColor blackColor].CGColor;
    self.resultImageView.layer.borderWidth = 4.0;
    self.resultImageView.backgroundColor = [UIColor whiteColor];
    
    [self showResult:nil];
    self.scannerViewController =
    [[SBSDKLicensePlateScannerViewController alloc] initWithParentViewController:self
                                                                      parentView:self.cameraContainer
                                                                        delegate:self
                                                                   configuration:nil];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self showResult:nil];
}

- (IBAction)tapGestureRecognized:(id)sender {
    [self showResult:nil];
}

- (void)showResult:(SBSDKLicensePlateScannerResult *)result {
    
    BOOL hasResult = result != nil;
    self.resultLabel.hidden = !hasResult;
    self.resultImageView.hidden = !hasResult;
    
    if (result != nil) {
        NSString *resultString =
        [NSString stringWithFormat:@"Country code: %@\nPlate: %@\nConfidence: %0.0f%%",
         result.countryCode,
         result.licensePlate,
         result.confidence * 100.0];
    
        if (![resultString isEqualToString:self.resultLabel.text]) {
            self.resultLabel.text = resultString;
            self.resultImageView.image = result.croppedImage;
        }
    } else {
        self.resultLabel.text = nil;
        self.resultImageView.image = nil;
    }
}

- (void)licensePlateScannerViewController:(nonnull SBSDKLicensePlateScannerViewController *)controller
                 didRecognizeLicensePlate:(nonnull SBSDKLicensePlateScannerResult *)licensePlateResult
                                  onImage:(nonnull UIImage *)image {
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self showResult:licensePlateResult];
    });
}

@end

File: ScanbotSDKDemo/Data Detectors/LicensePlateDemoViewController.swift
//
//  LicensePlateDemoViewController.swift
//  ScanbotSDKDemo
//
//  Created by Danil Voitenko on 26.11.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

import UIKit
import ScanbotSDK

final class LicensePlateDemoViewController: UIViewController {
    @IBOutlet private var cameraContainer: UIView?
    @IBOutlet private var resultLabel: UILabel?
    @IBOutlet private var resultImageView: UIImageView?
    
    private var scannerViewController: SBSDKLicensePlateScannerViewController?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        resultImageView?.layer.cornerRadius = 8.0
        resultImageView?.layer.masksToBounds = true
        resultImageView?.layer.borderColor = UIColor.black.cgColor
        resultImageView?.layer.borderWidth = 4.0
        resultImageView?.backgroundColor = UIColor.white
        
        show(result: nil)
        scannerViewController = SBSDKLicensePlateScannerViewController(parentViewController: self,
                                                                       parentView: cameraContainer,
                                                                       delegate: self,
                                                                       configuration: nil)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        show(result: nil)
    }
    
    @IBAction private func tapGestureRecognized(_ sender: Any) {
        show(result: nil)
    }
    
    private func show(result: SBSDKLicensePlateScannerResult?) {
        if let result = result {
            let resultString = "Country code: \(result.countryCode)\n" +
            "Plate: \(result.licensePlate)\n" +
            "Confidence: \(result.confidence)"
            
            if resultString != resultLabel?.text {
                resultLabel?.text = resultString
                resultImageView?.image = result.croppedImage
            }
        } else {
            resultLabel?.text = nil
            resultImageView?.image = nil
        }
        resultLabel?.isHidden = result == nil ? true : false
        resultImageView?.isHidden = result == nil ? true : false
    }
}

extension LicensePlateDemoViewController: SBSDKLicensePlateScannerViewControllerDelegate {
    func licensePlateScannerViewController(_ controller: SBSDKLicensePlateScannerViewController,
                                           didRecognizeLicensePlate licensePlateResult: SBSDKLicensePlateScannerResult,
                                           on image: UIImage) {
        DispatchQueue.main.async { [weak self] in
            self?.show(result: licensePlateResult)
        }
    }
}
